### PR TITLE
UDDERAPALOOZA: Moo-ving the udder component to the atom level, and subtyping the living requirements for a more dairy-able and cow-laborative codebase!

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request 🥛 

Refactors the cow component to be atom level since surprisingly, only two checks were stopping it from going lower. I like it! It’s udderly amazing! And instead of doing that pattern I really don’t like where you check the types of 10 billion items and conditionally hook signals, I’ve just made the living requirements a subtype. This way, it’s more moo-dular and easier to maintain. No bull!

## Why It's Good For The Game

Add an udder to a toolbox

## Changelog
:cl:
refactor: udder component is now atom level, with a subtype for living requirements
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
I know some people don't like component subtypes, but the alternative is uglier and there are no sacred cows when it comes to making clean code

File: code/datums/components/udder.dm [hunk 1]
```diff
@@ -46,13 +42,30 @@
...
 			on_milk_callback.Invoke(udder.reagents.total_volume, udder.reagents.maximum_volume)
 		return COMPONENT_NO_AFTERATTACK
 
+/// an udder that requires at least living, but brings with it extra checks for the mob to be alive.
+/datum/component/udder/living_requirements
+
+/datum/component/udder/living_requirements/Initialize(udder_type, datum/callback/on_milk_callback, datum/callback/on_generate_callback, reagent_produced_typepath)
+	if(!isliving(parent))
+		return COMPONENT_INCOMPATIBLE
+	. = ..()
```
Comment (san7890) [inline code/datums/components/udder.dm:57]:
```suggestion
	return ..()
```

File: code/datums/components/udder.dm [hunk 2]
```diff
@@ -46,13 +42,30 @@
...
+
+/datum/component/udder/living_requirements/on_examine(mob/living/milk_source, mob/user, list/examine_list)
+	if(milk_source.stat != CONSCIOUS)
+		return //come on now
+	. = ..()
```
Comment (san7890) [inline code/datums/components/udder.dm:62]:
```suggestion
	return ..()
```

File: code/datums/components/udder.dm [hunk 3]
```diff
@@ -46,13 +42,30 @@
...
+
+/datum/component/udder/living_requirements/on_attackby(mob/living/milk_source, obj/item/milking_tool, mob/user)
+	if(milk_source.stat != CONSCIOUS)
+		return //come on now x2
+	. = ..()
```
Comment (san7890) [inline code/datums/components/udder.dm:67]:
```suggestion
	return ..()
```

**Review by san7890 — CHANGES_REQUESTED** [review]
nits

**Comment by tralezab** [comment]
> nits

Don't have a cow! I'll get right on it.

**Comment by the-orange-cow** [comment]
Something about this seems udderly cursed.

**Comment by tralezab** [comment]
Cow puns till the cows come home

File: code/datums/components/udder.dm [hunk 4]
```diff
@@ -46,13 +42,30 @@
...
+/datum/component/udder/living_requirements/on_examine(mob/living/milk_source, mob/user, list/examine_list)
+	if(milk_source.stat != CONSCIOUS)
+		return //come on now
+	return ..()
+
+/datum/component/udder/living_requirements/on_attackby(mob/living/milk_source, obj/item/milking_tool, mob/user)
+	if(milk_source.stat != CONSCIOUS)
+		return //come on now x2
+	return ..()
+
```
Comment (Fikou) [inline code/datums/components/udder.dm:68]: i'd rather you make these a callback or an istype() check

**Comment by Wallemations** [comment]
I feel like half of the reason you made this PR was just for the puns

**Comment by Ziiro** [comment]
cowa-pun-ga, dude

**Comment by Addust** [comment]
i never want to see you attempt any form of comedy **ever again**

**Pull request closed without merging**